### PR TITLE
Make `PromptNode` and `PromptElement` types more React-like

### PR DESCRIPTION
**Pull request by unknown**
This PR makes `PromptNode` and `PromptElement`  types more similar to `React.ReactNode` and `React.ReactElement`, respectively. That is, the `PromptElement` type now encapsulates all the internal object representations of elements created through `Prompt.createElement` and JSX transpilation, and `PromptNode` now encapsulates all types that can be rendered (the `PromptElement` type in addition to literal types like strings and numbers).

Fixing these types allows some `Array.prototype.flat()` calls that were used to wrangle with Typescript to be removed.

NOTE: This change may require changing the return types of Priompt components from `PromptElement` to `PromptNode`. For example, the following would cause a type error because the `PromptElement` type does not include strings, even though `ExamplePrompt` is a renderable component.

```tsx
function ExamplePrompt(): PromptElement {
  return "Hello!";
}
```
Changing the `PromptElement` type to `PromptNode` would fix this type error.

### CODE REVIEW

File: priompt/src/lib.ts [hunk 1]
```diff
@@ -1561,9 +1559,8 @@ function computePriorityLevels(elem: AnyNode[] | AnyNode, parentPriority: number
 			levels.add(priority);
 			// we make the elem have this priority, so that we don't need to redo the priority calculation
 			elem.absolutePriority = priority;
-			// then for each child
-			for (const child of elem.children) {
-				computePriorityLevels(child, priority, levels);
+			if (Array.isArray(elem.children)) {
```
Comment (arvid220u) [inline priompt/src/lib.ts:1562]: hmm if elem.children isn't an array, shouldn't we still call computePriorityLevels on it? i may be missing something

File: priompt/src/lib.ts [hunk 2]
```diff
@@ -234,17 +230,19 @@ export function createElement(tag: ((props: BaseProps & Record<string, unknown>)
 	}
 }
 
-export function Fragment({ children }: { children: PromptElement[]; }): PromptElement {
-	// merge all the lists
-	return children.flat();
+export function Fragment({ children }: { children: PromptNode; }): PromptElement {
```
Comment (arvid220u) [inline priompt/src/lib.ts:233]: does Fragment need to return a PromptElement or is it fine to return a PromptNode? Ideally we wouldn't need to introduce the scope here

**Comment by arvid220u** [comment]
looks good otherwise i think!

in general a bit worried about refactors since every single part of our production code relies on priompt, and even well-intentioned and well-thought-through refactors have a risk of introducing bugs (sometimes just because the previous implementation was buggy and other code unintentionally relied on that specific bugginess!)

File: priompt/src/lib.ts (hunk 1 above)
Comment (unknown) [reply priompt/src/lib.ts:1562]: Good catch! I think this was a bug before as well then. Both `validateNotBothAbsoluteAndRelativePriority` and `validateNoChildrenHigherPriorityThanParent` have similarly buggy logic that also needs to be fixed.

File: priompt/src/lib.ts (hunk 2 above)
Comment (unknown) [reply priompt/src/lib.ts:233]: Ah I'll fix this as well. A `PromptNode` will work here.

**Comment by unknown** [comment]
> looks good otherwise i think!
> 
> in general a bit worried about refactors since every single part of our production code relies on priompt, and even well-intentioned and well-thought-through refactors have a risk of introducing bugs (sometimes just because the previous implementation was buggy and other code unintentionally relied on that specific bugginess!)

Understand this completely! I've tried to keep the changes as minimal as possible and focused on just making Priompt's element types more similar to React's. There aren't any significant logic changes other than removing `.flat()` from some children props and the bugs in `computePriorityLevels`, `validateNotBothAbsoluteAndRelativePriority`, and `validateNoChildrenHigherPriorityThanParent` you highlighted.

**Pull request closed without merging**